app/components/form_components/autocomplete/sort.js

```diff
@@ -0,0 +1,48 @@
+// Returns an array of integers representing the position of each regex match.
+const matchPositions = (string, regexes) => regexes.map(regex => string.search(regex)).filter(i => i >= 0)
+
+const calculateWeight = (name, rawSynonyms, query) => {
+  const regexes = query.split(/\s+/).map(word => new RegExp('\\b' + word, 'i'))
+  const synonyms = rawSynonyms || []
+
+  const nameMatchPositions = matchPositions(name, regexes)
+  const synonymMatchPositions = synonyms.map(synonym => matchPositions(synonym, regexes)).flat()
+
+  // Require all parts of the query to be matched:
+  if (nameMatchPositions.length !== regexes.length && synonymMatchPositions.length !== regexes.length) return 0
```

Comment:
Should this be `or` rather than `and`? or am I misreading it?

Reply:
hmm, I don't think this works. won't we need to check each word in the query to see if there was a particular match for it? We could have a query with two words: one with a name match and one with a synonym match which would break.

Reply:
also, if an option has multiple synonyms that start with the same letter you could easily match many synonyms for the same regex which would screw this up

Reply:
yep i need to re-think this. i don't think i'm 100% sure what we want to happen.

check: `nameMatchPositions.length !== regexes.length && synonymMatchPositions.length !== regexes.length`

query: "master arts"
name: "master"
synonym: "something"
**outcome**: `true` && `true` => no match (correct)

query: "master arts"
name: "something"
synonym: "master"
**outcome**: `true` && `true` => no match (correct)

query: "master arts"
name: "master"
synonym: "arts"
**outcome**: `true` && `true` => no match (correct)

query: "master arts"
name: "master"
synonym: "master arts"
**outcome**: `true` && `false` => no match (incorrect)

Reply:
Updated logic and added tests to cover what happens in different situations.